### DISCUSSION TITLE
Watertightness of Mesh with an Edge Shared by 4 Faces

### DISCUSSION THREAD

**Discussion by Eric-Vin** [elalish/manifold#773] [Q&A] [answered]
I'm working on converting a representation of a grid of Voxels to an equivalent Mesh representation in Trimesh and I had questions about Manifolds definition of a valid mesh, specifically with regards to watertightness, since I'm using it as an engine in Trimesh. The case in question is one where an edge has 4 faces connected to it (example in [this](https://github.com/mikedh/trimesh/issues/2165) issue), and it was unclear to me if such a mesh would be considered valid/watertight by Manifold. Thanks for your help!

**Comment by elalish** [comment, marked as answer]
Indeed, we use oriented 2-manifolds (see our manifoldness definition on the [wiki](https://github.com/elalish/manifold/wiki/Manifold-Library#manifoldness)), so 4-manifolds are no good. However, when creating a `Manifold` from a `Mesh` or a `MeshGL`, we do fix some basic manifoldness problems automatically. 4-manifold edges and pinched verts *should* be fixable, though I wouldn't guarantee it. We fix these cases by duplicating verts - so your meshes will either magically work, or you'll need to do some similar vertex duplication until you get down to a 2-manifold. Basically you just have to pick which pairs of faces are actually topological neighbors and which are not.

**Reply by Eric-Vin** [reply]
Thanks for the quick response! Duplicating vertices had come up on the Trimesh issue I mentioned, but I was worried they might be categorized as self intersections and break the guarantees Manifold offers. If duplicate vertices are OK for, I'll use that strategy. Thanks so much again!

**Reply by elalish** [reply]
Yes, duplicating verts is totally fine in Manifold - it's the only way we can keep our Boolean ops manifold. We are quite tolerant of self-intersection, hence our concept of &epsilon;-valid. Everything should continue to work, but if your self-intersections exceed your shape's `precision`, you'll eventually get ugly (though manifold) results.

**Reply by Eric-Vin** [reply]
Fantastic, thanks so much!
